src/components/button/button.js

```diff
@@ -1,33 +1,99 @@
 const React = require('react');
+
+const Control = require('../control/control');
 const provide = require('../../provider/provider');
 
-module.exports = class Button extends React.Component {
+module.exports = class Button extends Control {
+
+    constructor(props) {
+        super(props);
+
+        this._focusedByPointer = false;
+        this._isPointerPressInProgress = false;
+    }
+
+    _focus() {
+        if(this._isPointerPressInProgress) return;
+
+        if(this._focusedByPointer) {
+            this.setStateAndMod({focused: true});
+        } else {
+            this.setStateAndMod({'focused-hard': true});
+        }
+    }
+
+    _onFocus(e) {
+        this._focus();
+
+        this.props.onFocus && this.props.onFocus(e, this.state);
+    }
+
+    _onBlur(e) {
+        this.setStateAndMod({pressed: false, 'focused-hard': false});
+
+        super._onBlur(e);
+    }
+
+    _onClick(e) {
+        this.props.onClick && this.onClick(e, this.state);
```

Comment:
it's not good to pass inner state of component somewhere in callback. What's the point of it?